File: cli/manage-accounts.ts
import fs from 'node:fs';
import path from 'node:path';
import readline from 'node:readline';

const DATA_DIR = path.join(process.cwd(), 'data');
const ACCOUNTS_FILE = path.join(DATA_DIR, 'accounts.json');

interface Account {
  id: string;
  label: string;
  credentials: any;
}

function loadAccounts(): Account[] {
  if (!fs.existsSync(ACCOUNTS_FILE)) {
    return [];
  }
  const content = fs.readFileSync(ACCOUNTS_FILE, 'utf-8');
  try {
    return JSON.parse(content);
  } catch (e) {
    console.error('Failed to parse accounts.json:', e);
    return [];
  }
}

function saveAccounts(accounts: Account[]) {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
  fs.writeFileSync(ACCOUNTS_FILE, JSON.stringify(accounts, null, 2), { mode: 0o600 });
}

async function ask(question: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

async function askMultiLine(question: string): Promise<string> {
  console.log(question);
  console.log('(入力を終了するには Ctrl+D を押してください)');
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  
  return new Promise((resolve) => {
    let lines: string[] = [];
    rl.on('line', (line) => {
      lines.push(line);
    });
    rl.on('close', () => {
      resolve(lines.join('\n'));
    });
  });
}

async function addAccount() {
  console.log('\n=== Add Gemini Account ===\n');
  const label = await ask('Enter a label for this account (e.g., user@gmail.com): ');
  if (!label) {
    console.log('❌ Label is required.');
    return;
  }

  const credsRaw = await askMultiLine('Paste your credentials (content of oauth_creds.json):\n');
  
  let credentials;
  try {
    credentials = JSON.parse(credsRaw);
  } catch (e) {
    console.error('❌ Invalid JSON format. Please check your credentials and try again.');
    return;
  }

  const accounts = loadAccounts();
  const id = `account-${accounts.length}-${Date.now().toString(36)}`;
  
  accounts.push({ id, label, credentials });
  saveAccounts(accounts);

  console.log(`\n✅ Account "${label}" added successfully (ID: ${id})`);
  console.log(`Current account count: ${accounts.length}`);
}

function listAccounts() {
  const accounts = loadAccounts();
  if (accounts.length === 0) {
    console.log('No accounts registered.');
    return;
  }

  console.log('\n=== Registered Accounts ===\n');
  console.table(accounts.map(a => ({
    ID: a.id,
    Label: a.label,
    Credentials: '*** (MASKED) ***'
  })));
}

async function removeAccount() {
  const accounts = loadAccounts();
  if (accounts.length === 0) {
    console.log('No accounts to remove.');
    return;
  }

  listAccounts();
  const idToRemove = await ask('\nEnter the ID of the account to remove (or press Enter to cancel): ');
  if (!idToRemove) return;

  const filtered = accounts.filter(a => a.id !== idToRemove);
  if (filtered.length === accounts.length) {
    console.log(`❌ Account with ID "${idToRemove}" not found.`);
    return;
  }

  saveAccounts(filtered);
  console.log(`\n✅ Account "${idToRemove}" has been removed.`);
}

async function main() {
  const command = process.argv[2];

  switch (command) {
    case 'add':
      await addAccount();
      break;
    case 'list':
      listAccounts();
      break;
    case 'remove':
      await removeAccount();
      break;
    default:
      console.log('Usage: npm run account:[add|list|remove]');
  }
}

main().catch(console.error);

File: src/claude2gemini/cli/manage_accounts.py
"""CLI for managing Gemini API accounts."""

import json
import os
import sys

DATA_DIR = os.path.join(os.getcwd(), "data")
ACCOUNTS_FILE = os.path.join(DATA_DIR, "accounts.json")


def _load_accounts() -> list[dict]:
    if not os.path.exists(ACCOUNTS_FILE):
        return []
    with open(ACCOUNTS_FILE) as f:
        data = json.load(f)
    if isinstance(data, list):
        return data
    return []


def _save_accounts(accounts: list[dict]) -> None:
    os.makedirs(DATA_DIR, exist_ok=True)
    with open(ACCOUNTS_FILE, "w") as f:
        json.dump(accounts, f, indent=2)


def cmd_list() -> None:
    """List all registered accounts."""
    accounts = _load_accounts()
    if not accounts:
        print("No accounts registered.")
        return
    print(f"\nRegistered Accounts ({len(accounts)}):\n")
    for a in accounts:
        print(f"  ID: {a.get('id', '?')}")
        print(f"  Label: {a.get('label', '?')}")
        print(f"  Credentials: *** (MASKED) ***")
        print()


def cmd_add(label: str, creds_json: str) -> None:
    """Add a new account."""
    try:
        credentials = json.loads(creds_json)
    except json.JSONDecodeError:
        print("Error: Invalid JSON format for credentials.")
        sys.exit(1)

    accounts = _load_accounts()
    account_id = f"account-{len(accounts)}-{os.urandom(4).hex()}"
    accounts.append({"id": account_id, "label": label, "credentials": credentials})
    _save_accounts(accounts)
    print(f"Account '{label}' added successfully (ID: {account_id})")


def cmd_remove(account_id: str) -> None:
    """Remove an account by ID."""
    accounts = _load_accounts()
    before = len(accounts)
    accounts = [a for a in accounts if a.get("id") != account_id]
    if len(accounts) == before:
        print(f"Account with ID '{account_id}' not found.")
        return
    _save_accounts(accounts)
    print(f"Account '{account_id}' removed.")


def main() -> None:
    command = sys.argv[1] if len(sys.argv) > 1 else None

    if command == "list":
        cmd_list()
    elif command == "add":
        if len(sys.argv) < 4:
            print("Usage: python -m claude2gemini.cli.manage_accounts add <label> <credentials_json>")
            sys.exit(1)
        cmd_add(sys.argv[2], sys.argv[3])
    elif command == "remove":
        if len(sys.argv) < 3:
            print("Usage: python -m claude2gemini.cli.manage_accounts remove <id>")
            sys.exit(1)
        cmd_remove(sys.argv[2])
    else:
        print("Usage: python -m claude2gemini.cli.manage_accounts [add|list|remove]")
        sys.exit(1)


if __name__ == "__main__":
    main()
